File: examples/canonical-aws-lambda/s3-trigger-image-moderation/python/tests/smoke/test_handler.py
import pathlib
import sys
import unittest

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parents[2] / "src"))

from handler import Clients, DuplicateObjectVersion, lambda_handler


class Table:
    def __init__(self):
        self.items = {}

    def put_item(self, **kwargs):
        item = kwargs["Item"]
        pk = item["pk"]
        if kwargs.get("ConditionExpression") and pk in self.items:
            raise DuplicateObjectVersion()
        self.items[pk] = item


class Rekognition:
    def detect_labels(self, **kwargs):
        return {"Labels": [{"Name": "Document"}]}

    def detect_moderation_labels(self, **kwargs):
        return {"ModerationLabels": [{"Name": "Explicit Nudity"}]}


class Events:
    def __init__(self):
        self.entries = []

    def put_events(self, **kwargs):
        self.entries.extend(kwargs["Entries"])


class HandlerSmokeTest(unittest.TestCase):
    def test_flags_image_and_emits_event(self):
        table = Table()
        events = Events()
        result = lambda_handler(event("v1"), clients=Clients(table, Rekognition(), events))
        self.assertEqual(result["decision"], "FLAG")
        self.assertEqual(len(table.items), 1)
        self.assertEqual(len(events.entries), 1)


def event(version):
    return {"Records": [{"s3": {"bucket": {"name": "images"}, "object": {"key": "a.jpg", "versionId": version}}}]}


if __name__ == "__main__":
    unittest.main()

File: examples/canonical-aws-lambda/s3-trigger-image-moderation/typescript/tests/smoke/test_handler.js
const assert = require("assert");
const { handler } = require("../../src/handler.js");

async function main() {
  const claimed = new Set();
  const flagged = [];
  const result = await handler(event("v1"), {
    claimObjectVersion: async (key) => {
      if (claimed.has(key)) return false;
      claimed.add(key);
      return true;
    },
    detectLabels: async () => [{ Name: "Document" }],
    detectModerationLabels: async () => [{ Name: "Explicit Nudity" }],
    recordDecision: async () => {},
    publishFlagged: async (detail) => flagged.push(detail)
  });
  assert.equal(result.decision, "FLAG");
  assert.equal(claimed.size, 1);
  assert.equal(flagged.length, 1);
}

function event(version) {
  return { Records: [{ s3: { bucket: { name: "images" }, object: { key: "a.jpg", versionId: version } } }] };
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
